src/event/ngx_event_openssl_cache.c

```diff
@@ -180,18 +284,122 @@ ngx_ssl_cache_fetch(ngx_conf_t *cf, ngx_uint_t index, char **err,
 
 
 void *
-ngx_ssl_cache_connection_fetch(ngx_pool_t *pool, ngx_uint_t index, char **err,
-    ngx_str_t *path, void *data)
+ngx_ssl_cache_connection_fetch(ngx_ssl_cache_t *cache, ngx_pool_t *pool,
+    ngx_uint_t index, char **err, ngx_str_t *path, void *data)
 {
-    ngx_ssl_cache_key_t  id;
+    void                  *value;
+    time_t                 now;
+    uint32_t               hash;
+    ngx_file_info_t        fi;
+    ngx_ssl_cache_key_t    id;
+    ngx_ssl_cache_type_t  *type;
+    ngx_ssl_cache_node_t  *cn;
 
     *err = NULL;
 
     if (ngx_ssl_cache_init_key(pool, index, path, &id) != NGX_OK) {
         return NULL;
     }
 
-    return ngx_ssl_cache_types[index].create(&id, err, data);
+    type = &ngx_ssl_cache_types[index];
+
+    if (cache == NULL) {
+        return type->create(&id, err, &data);
+    }
+
+    now = ngx_time();
+
+    hash = ngx_murmur_hash2(id.data, id.len);
+
+    cn = ngx_ssl_cache_lookup(cache, type, &id, hash);
+
+    if (cn != NULL) {
+        ngx_queue_remove(&cn->queue);
+
+        if (now - cn->created > cache->valid) {
+
+            if (ngx_file_info(cn->id.data, &fi) == NGX_FILE_ERROR
+                || ngx_file_uniq(&fi) != cn->uniq
+                || ngx_file_mtime(&fi) != cn->mtime)
+            {
+                ngx_log_debug1(NGX_LOG_DEBUG_CORE, pool->log, 0,
+                               "cached ssl file changed: %s", cn->id.data);
+
+                type->free(cn->value);
+
+                cn->value = type->create(&id, err, &data);
+                if (cn->value == NULL) {
+                    ngx_rbtree_delete(&cache->rbtree, &cn->node);
+
+                    cache->current--;
+
+                    ngx_free(cn);
+
+                    return NULL;
+                }
+
+                if (ngx_file_info(cn->id.data, &fi) != NGX_FILE_ERROR) {
+                    cn->mtime = ngx_file_mtime(&fi);
+                    cn->uniq = ngx_file_uniq(&fi);
+                }
+            }
+
+            cn->created = now;
+        }
+
+        goto found;
+    }
+
+    value = type->create(&id, err, &data);
+    if (value == NULL) {
+        return NULL;
+    }
+
+    if (data == NGX_SSL_CACHE_DISABLED) {
+        return value;
+    }
+
+    cn = ngx_alloc(sizeof(ngx_ssl_cache_node_t) + id.len + 1, pool->log);
+    if (cn == NULL) {
+        type->free(value);
+        return NULL;
+    }
+
+    cn->node.key = hash;
+    cn->id.data = (u_char *)(cn + 1);
+    cn->id.len = id.len;
+    cn->id.type = id.type;
+    cn->type = type;
+    cn->created = now;
+    cn->value = value;
+
+    if (ngx_file_info(cn->id.data, &fi) != NGX_FILE_ERROR) {
```

Comment:
1. ngx_file_info can be skipped for non-file cache entries
2. We haven't copied the `id.data` yet

```suggestion
    if (id.type == NGX_SSL_CACHE_PATH
        && ngx_file_info(id.data, &fi) != NGX_FILE_ERROR)
    {
```

Reply:
Fixed, thanks.